s3auth-relay/src/main/java/com/s3auth/relay/FtpFacade.java

```diff
@@ -0,0 +1,247 @@
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.s3auth.relay;
+
+import com.jcabi.aspects.Loggable;
+import com.jcabi.aspects.Tv;
+import com.jcabi.log.Logger;
+import com.jcabi.log.VerboseRunnable;
+import com.jcabi.log.VerboseThreads;
+import com.s3auth.hosts.Hosts;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import javax.validation.constraints.NotNull;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+import org.apache.commons.net.ftp.FTPReply;
+
+/**
+ * FTP facade (port listener).
+ *
+ * <p>The class is instantiated in {@link Main}, once per application run.
+ *
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Felipe Pina (felipe.pina@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ * @see Main
+ * @checkstyle ClassDataAbstractionCoupling (10 lines)
+ * @todo #213:30min Implement TLS secure port listening in a manner analogous to
+ *  HttpFacade.
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server" })
+@SuppressWarnings("PMD.DoNotUseThreads")
+@Loggable(Loggable.DEBUG)
+final class FtpFacade implements Closeable {
+
+    /**
+     * How many threads to use.
+     */
+    private static final int THREADS = Tv.HUNDRED;
+
+    /**
+     * Executor service, with socket openers.
+     */
+    private final transient ScheduledExecutorService frontend =
+        Executors.newScheduledThreadPool(2, new VerboseThreads("FTP-front"));
+
+    /**
+     * Executor service, with consuming threads.
+     */
+    private final transient ScheduledExecutorService backend =
+        Executors.newScheduledThreadPool(
+            FtpFacade.THREADS,
+            new VerboseThreads("FTP-back")
+        );
+
+    /**
+     * Blocking queue of ready-to-be-processed sockets.
+     */
+    private final transient BlockingQueue<Socket> sockets =
+        new SynchronousQueue<Socket>();
+
+    /**
+     * Server socket.
+     */
+    private final transient ServerSocket server;
+
+    /**
+     * Public ctor.
+     * @param hosts Hosts
+     * @param port Port number
+     * @throws IOException If can't initialize
+     */
+    FtpFacade(@NotNull final Hosts hosts, final int port)
+        throws IOException {
+        this.server = new ServerSocket(port);
+        final FtpThread thread = new FtpThread(this.sockets, hosts);
+        final Runnable runnable = new VerboseRunnable(
+            new FtpFacade.FTPThreadRunnable(thread), true, true
```

Comment:
why not create an embedded class instead? would be much less verbose. now it's confusing and the name of the class is rather weird